# Intro to Pandas

Where NumPy uses arrays, Pandas uses something called a "dataframe".  They're more flexible than NumPy arrays.  All that mess we had to do with the labels and state names in NumPy?  Well.

Pandas supports labels for columns.  They can be strings. 

Pandas columns can be of different types.  All of a column should be of the same type, but what if we have a string label for each row, then dates in one column and counts (integers) in another? 

So Pandas dataframes are more "spreadsheety".




<img src="images/pic-of-pandas-dataframe.png" />

Pandas is very complex and large.  It can do A LOT of stuff.  Let's just scratch some surface.  

* Check out the [User Guide](http://pandas-docs.github.io/pandas-docs-travis/user_guide/index.html)  if you want to be overwhelmed. 

* The [cookbook](http://pandas-docs.github.io/pandas-docs-travis/user_guide/cookbook.html) is actually a good place to go after this lab.  

Let's load up some data!



In [1]:
import pandas as pd
# Pandas is usually imported as "pd" 
# and there is a BETTER way to load csv data directly into Pandas.

#reads into dataframe, index_col = titles are in the 0th column
f500 = pd.read_csv('fortune500data2017.csv', index_col=0)
f500.head(4)

rank  rank_prev  Revenue_M  Revenue_Change  \
title                                                                  
Walmart                      1          1     485873             0.8   
State Grid                   2          2     315199            -4.4   
Sinopec Group                3          4     267518            -9.1   
China National Petroleum     4          3     262573           -12.3   

                          Profits_M  Assets_M  Profit_Change  Employees  \
title                                                                     
Walmart                     13643.0    198825           -7.2    2300000   
State Grid                   9571.3    489838           -6.2     926067   
Sinopec Group                1257.9    310726          -65.0     713288   
China National Petroleum     1867.5    585619          -73.7    1512048   

                                          ceo               industry  \
title                                                                  
Walmart                   C. Douglas McMillon  General Merchandisers   
State Grid                            Kou Wei              Utilities   
Sinopec Group                       Wang Yupu     Petroleum Refining   
China National Petroleum        Zhang Jianhua     Petroleum Refining   

                             sector         location                 website  \
title                                                                          
Walmart                   Retailing  Bentonville, AR  http://www.walmart.com   
State Grid                   Energy   Beijing, China  http://www.sgcc.com.cn   
Sinopec Group                Energy   Beijing, China  http://www.sinopec.com   
China National Petroleum     Energy   Beijing, China  http://www.cnpc.com.cn   

                          Years_on_list  Stock_Equity  profit_as_pct_revenue  \
title                                                                          
Walmart                              23         77798                    2.8   
State Grid                           17        209456                    3.0   
Sinopec Group                        19        106523                    0.5   
China National Petroleum             17        301893                    0.7   

                          profit_as_pct_assets  profit_as_pct_equity  
title                                                                 
Walmart                                    6.9                  17.5  
State Grid                                 2.0                   4.6  
Sinopec Group                              0.4                   1.2  
China National Petroleum                   0.3                   0.6

See the `index_col` parameter?  That tells Pandas that column 0 contains index labels for the data.  Row 1 is assumed to be header data (strings) usually.  Lots of options you can check in the docs.

See the head method on the f500 object?  That'll show the first four rows.  Useful if there's, like, 19 billion rows.  There's also a .tail method if you want the end ones.  Nice to check the data for completeness.

You can also specify the index_col by the name in the header.

#### --> 1.  In the cell below
* Change index_col to be 'title' instead of 0
* Show the last 3 rows with tail()

**Double click for answer**
<!--

f500 = pd.read_csv('fortune500data2017.csv', index_col='title')
-->


In [4]:


f500 = pd.read_csv('fortune500data2017.csv', index_col = 'title')
f500.tail(3)

rank  rank_prev  Revenue_M  Revenue_Change  \
title                                                                   
Wm. Morrison Supermarkets   498        437      21741           -11.3   
TUI                         499        467      21655            -5.5   
AutoNation                  500          0      21609             3.6   

                           Profits_M  Assets_M  Profit_Change  Employees  \
title                                                                      
Wm. Morrison Supermarkets      406.4     11630           20.4      77210   
TUI                           1151.7     16247          195.5      66779   
AutoNation                     430.5     10060           -2.7      26000   

                                          ceo              industry  \
title                                                                 
Wm. Morrison Supermarkets      David T. Potts  Food and Drug Stores   
TUI                         Friedrich Joussen       Travel Services   
AutoNation                 Michael J. Jackson   Specialty Retailers   

                                       sector             location  \
title                                                                
Wm. Morrison Supermarkets  Food & Drug Stores    Bradford, Britain   
TUI                         Business Services     Hanover, Germany   
AutoNation                          Retailing  Fort Lauderdale, FL   

                                             website  Years_on_list  \
title                                                                 
Wm. Morrison Supermarkets   http://www.morrisons.com             13   
TUI                          http://www.tuigroup.com             23   
AutoNation                 http://www.autonation.com             12   

                           Stock_Equity  profit_as_pct_revenue  \
title                                                            
Wm. Morrison Supermarkets          5111                    1.9   
TUI                                3006                    5.3   
AutoNation                         2310                    2.0   

                           profit_as_pct_assets  profit_as_pct_equity  
title                                                                  
Wm. Morrison Supermarkets                   3.5                   8.0  
TUI                                         7.1                  38.3  
AutoNation                                  4.3                  18.6

`f500` is the variable that holds our dataframe.  There is **a lot** of stuff in there.

#### --> 2.  Begin to inspect the dataframe.

Inspect the types of the f500 dataframe using the `dtypes` property.  This is a property, not a method. So don't put () on the end (which is used to call methods and functions).  f500.dtypes  

Try… info method.
In the following cell, make a call to f500.info() - this function PRINTS stuff.  You can't assign it to a variable.  Pandas is used for exploration so there are a lot of tools for doing just that.

**Double Click Here for answer**
<!--
print("Dtypes:")
# The dtypes property will show the types of the columns by name.
# What type do the string columns show up as? fill in the blank: object


print("\n\ninfo()")
print(f500.info())

-->



In [9]:
print("Dtypes:")
print(f500.dtypes)

# The dtypes property will show the types of the columns by name.
# What type do the string columns show up as? fill in the blank: _____________

print("\n\ninfo():")
print(f500.info())

Dtypes:
rank                       int64
rank_prev                  int64
Revenue_M                  int64
Revenue_Change           float64
Profits_M                float64
Assets_M                   int64
Profit_Change            float64
Employees                  int64
ceo                       object
industry                  object
sector                    object
location                  object
website                   object
Years_on_list              int64
Stock_Equity               int64
profit_as_pct_revenue    float64
profit_as_pct_assets     float64
profit_as_pct_equity     float64
dtype: object


info():
<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, Walmart to AutoNation
Data columns (total 18 columns):
rank                     500 non-null int64
rank_prev                500 non-null int64
Revenue_M                500 non-null int64
Revenue_Change           498 non-null float64
Profits_M                499 non-null float64
Assets_M                 500 non-null

Segway:
    `int64` is a 64 bit integer.  Most computers these days have 64 bit CPUs.

    * the integer is signed, and one bit is used for the sign bit (sort of right - the number is actually in twos compliment notation
    * so thats +/- 2 ^ 63  whcih is roughly +- 10 ^ 19  .... because 2^10 is roughly 10^3
    * That's a very large integer range.
    * The first microprocessors were 8-bit processors.

## Indexing

One cool thing about Pandas is that you can use the column index names and row index names to index into the data and not just index numbers like with NumPy.

You do this through the `loc` property.  It provides a reference into the data 
* based on name or position.
* for rows and / or columns

#### --> 3.  Try:

Below, 
slice out the 'Years_on_list' column with: `years = f500.loc[:, 'Years_on_list']`
Print the shape of the years variable  with the `shape` property.  
And show the type of it.
After that on one line put `years` to make the JN console dump out that variable.


**Double Click Here for answer**
<!--
print(years.shape, type(years))

years = f500.loc[:, 'Years_on_list']

print(years.shape)
print(type(years))
years
# What type is the years variable?  Series
-->



In [15]:
years = f500.loc[:, 'Years_on_list']  # the colon is a "slice" of all rows.
print(years.shape)
type(years)

# What type is the years variable?  a series!

(500,)


pandas.core.series.Series

# Pandas Series and Slicing

## Pandas Series
Note that you have a "series" datatype instead of a dataframe.  A series is the other fundamental Pandas type.  It's analogous to a vector in Numpy.  One dimensional.  Notice that the name of the company is not included. 

* A Pandas dataframe (df for short) is a collection of series (columns).

* Each series has an embedded NumPy array in it (accessible via its values property)




#### --> 4.  More exploring.

Try displaying years with print(years) to see what that looks like.

A dataframe (df)  is basically a list of series actually. Each column of a dataframe is a series.  And the loc property with its indexing supports selecting arbitrary columns.

Try selecting two columns instead of one:
```python
    rankAndProfits = f500.loc[:, ['rank', 'Profits_M']]
```
And print out the type...
    `type(rankAndProfits)`

And then lastly on one line to display the result nicely...
    `rankAndProfits`
Notice that JN knows how to display that nicely. Much nicer than `print`


**Double Click Here for answer**
<!--
rankAndProfits = f500.loc[:, ['rank', 'Profits_M']]
print("It's of type: ", type(rankAndProfits))
rankAndProfits
-->

In [16]:
rankAndProfits = f500.loc[:, ['rank', 'Profits_M']]
print("It's of type: ", type(rankAndProfits))
rankAndProfits

It's of type:  <class 'pandas.core.frame.DataFrame'>


rank  Profits_M
title                                          
Walmart                            1    13643.0
State Grid                         2     9571.3
Sinopec Group                      3     1257.9
China National Petroleum           4     1867.5
Toyota Motor                       5    16899.3
...                              ...        ...
Teva Pharmaceutical Industries   496      329.0
New China Life Insurance         497      743.9
Wm. Morrison Supermarkets        498      406.4
TUI                              499     1151.7
AutoNation                       500      430.5

[500 rows x 2 columns]

## You can index directly into a df to get a series or a smaller df.

If  you want all rows you actually don't need to use the `loc` property.

You can index into the df `f500` directly.

#### --> 5.  Indexing into a df directly and carving it up.

* Try indexing directly into f500 with the string 'rank' 

* And then in another cell index into f500 with a list (this is kind of weird)
    `['rank','Profits_M']`


**Double Click Here for answer**
<!--
# Index into f500 with string 'rank'
f500['rank']

## what type do you THINK that is? type(f500['rank']) is a Series

-----
f500[  ['rank','Profits_M']  ]
## What type do you THINK that is?  It's a new dataframe.

Dataframes have more than one column. One column by itself is a series.
-->


In [20]:
# Index into f500 with string 'rank'
rank = f500['rank']
type(rank)

### what type do you THINK that is? a series!

pandas.core.series.Series

In [21]:
# Index into f500 with a LIST .... ['rank','Profits_M'] 
wrank = f500[['rank', 'Profits_M']]
type(wrank)

### What type do you think that is? a DataFrame!

pandas.core.frame.DataFrame

### selecting rows


You can also select several rows by name with `.loc` property.  And you can combine them. 

#### -->  6.  Try in 
```python
# select a few rows:
aFewCompanies = f500.loc[["Apple", "Walmart", "IBM"]]
#and this one subselects columns and then a few rows.
aFewCompaniesAndCols = f500[["rank","Profits_M"]].loc[["Apple", "Walmart", "IBM"]]




In [22]:
aFewCompanies = f500.loc[["Apple", "Walmart", "IBM"]]
aFewCompanies

rank  rank_prev  Revenue_M  Revenue_Change  Profits_M  Assets_M  \
title                                                                      
Apple       9          9     215639            -7.7    45687.0    321686   
Walmart     1          1     485873             0.8    13643.0    198825   
IBM        81         82      79919            -3.1    11872.0    117470   

         Profit_Change  Employees                  ceo  \
title                                                    
Apple            -14.4     116000      Timothy D. Cook   
Walmart           -7.2    2300000  C. Douglas McMillon   
IBM              -10.0     414400  Virginia M. Rometty   

                                industry      sector         location  \
title                                                                   
Apple        Computers, Office Equipment  Technology    Cupertino, CA   
Walmart            General Merchandisers   Retailing  Bentonville, AR   
IBM      Information Technology Services  Technology       Armonk, NY   

                        website  Years_on_list  Stock_Equity  \
title                                                          
Apple      http://www.apple.com             15        128249   
Walmart  http://www.walmart.com             23         77798   
IBM          http://www.ibm.com             23         18246   

         profit_as_pct_revenue  profit_as_pct_assets  profit_as_pct_equity  
title                                                                       
Apple                     21.2                  14.2                  35.6  
Walmart                    2.8                   6.9                  17.5  
IBM                       14.9                  10.1                  65.1

In [23]:
aFewCompaniesAndCols = f500[["rank","Profits_M"]].loc[["Apple", "Walmart", "IBM"]]
aFewCompaniesAndCols

rank  Profits_M
title                   
Apple       9    45687.0
Walmart     1    13643.0
IBM        81    11872.0


### Direct indexing is designed primarily for convenience.

**You can slice a df by rows (you can't slice by columns**

Slice by rows is the most common (which is why you can't slice by columns):


So 
* if you put ONE string in the index, you get the column of that name
* If you put a slice, you get rows.
* If you put a list of things, you get the columns that have those names.

#### -->  7. Try slicing to understand it...

* `f500['rank']` to pull out a single column
* `f500.loc['Walmart']` to pull out a single row.
* `f500["Walmart":"Toyota Motor"]`  # all rows from Walmart to Toyota Motor **inclusive** (unlike Python slice which does not include the end)
* `f500["Walmart":"Toyota Motor"][["rank","Profits_M"]]` to slice through the top rows, and then from there, select out those two columns rank and Profits_M.






In [24]:
#  single column 'rank'

f500['rank']



title
Walmart                             1
State Grid                          2
Sinopec Group                       3
China National Petroleum            4
Toyota Motor                        5
                                 ... 
Teva Pharmaceutical Industries    496
New China Life Insurance          497
Wm. Morrison Supermarkets         498
TUI                               499
AutoNation                        500
Name: rank, Length: 500, dtype: int64

In [26]:
f500.loc["Walmart"]  # single row?  No.  Use loc.

rank                                          1
rank_prev                                     1
Revenue_M                                485873
Revenue_Change                              0.8
Profits_M                                 13643
Assets_M                                 198825
Profit_Change                              -7.2
Employees                               2300000
ceo                         C. Douglas McMillon
industry                  General Merchandisers
sector                                Retailing
location                        Bentonville, AR
website                  http://www.walmart.com
Years_on_list                                23
Stock_Equity                              77798
profit_as_pct_revenue                       2.8
profit_as_pct_assets                        6.9
profit_as_pct_equity                       17.5
Name: Walmart, dtype: object

In [28]:
# Use a slice to pull all rows walmart through toyota.  Try not to capypasta!
f500.loc["Walmart":"Toyota Motor"]



rank  Profits_M
title                                    
Walmart                      1    13643.0
State Grid                   2     9571.3
Sinopec Group                3     1257.9
China National Petroleum     4     1867.5
Toyota Motor                 5    16899.3

In [11]:
# rank and Profits_M columns for rows Walmart through Toyota Motor
f500.loc["Walmart":"Toyota Motor"][["rank", "Profits_M"]]

rank  Profits_M
title                                    
Walmart                      1    13643.0
State Grid                   2     9571.3
Sinopec Group                3     1257.9
China National Petroleum     4     1867.5
Toyota Motor                 5    16899.3

# Selecting Summary

single_col = `f500["rank"]`
single_row = `f500.loc["Walmart"]`
multiple_cols = `f500[["rank", 'ceo']]`
multiple_rows = `f500.loc[['Apple','IBM']]`
span_of_rows = `f500['Apple':'IBM']`


In [29]:
multiple_cols = f500[["rank", 'ceo']]

In [30]:
multiple_rows = f500.loc[['Apple','IBM']]

## Describing Series and DFs
Like the `df.info()` method, the `df.describe()` method is great for peeking at data.

There are two forms of it.

#### --> 8. Describe  - experiment in two cells below:

* One that describes the numeric data.<br />
   Try  `f500.describe()`

* One that describes the object data:<br />
    Try:  `f500.describe(include=['O']) `<br />
    that's an OH, not a ZERO - it represents that you want OBJECT data.



In [31]:
# Describe the numeric parts of the f500 df
f500.describe()

rank   rank_prev      Revenue_M  Revenue_Change     Profits_M  \
count  500.000000  500.000000     500.000000      498.000000    499.000000   
mean   250.500000  222.134000   55416.358000        4.538353   3055.203206   
std    144.481833  146.941961   45725.478963       28.549067   5171.981071   
min      1.000000    0.000000   21609.000000      -67.300000 -13038.000000   
25%    125.750000   92.750000   29003.000000       -5.900000    556.950000   
50%    250.500000  219.500000   40236.000000        0.550000   1761.600000   
75%    375.250000  347.250000   63926.750000        6.975000   3954.000000   
max    500.000000  500.000000  485873.000000      442.300000  45687.000000   

           Assets_M  Profit_Change     Employees  Years_on_list  \
count  5.000000e+02     436.000000  5.000000e+02     500.000000   
mean   2.436323e+05      24.152752  1.339983e+05      15.036000   
std    4.851937e+05     437.509566  1.700878e+05       7.932752   
min    3.717000e+03    -793.700000  3.280000e+02       1.000000   
25%    3.658850e+04     -22.775000  4.293250e+04       7.000000   
50%    7.326150e+04      -0.350000  9.291050e+04      17.000000   
75%    1.805640e+05      17.700000  1.689172e+05      23.000000   
max    3.473238e+06    8909.500000  2.300000e+06      23.000000   

        Stock_Equity  profit_as_pct_revenue  profit_as_pct_assets  \
count     500.000000             493.000000            487.000000   
mean    30628.076000               5.652130              3.115606   
std     43642.576833               7.790644              4.401207   
min    -59909.000000             -40.100000            -23.700000   
25%      7553.750000               1.400000              0.700000   
50%     15809.500000               3.800000              2.300000   
75%     37828.500000               8.100000              4.700000   
max    301893.000000              46.100000             23.700000   

       profit_as_pct_equity  
count            488.000000  
mean              16.684221  
std               52.896765  
min             -160.300000  
25%                5.900000  
50%               10.000000  
75%               15.800000  
max              847.200000

In [33]:
# describe the object parts of the f500 df so.. the things which are not numbers

f500.describe(include=['O'])

ceo                       industry      sector  \
count               500                            500         500   
unique              500                             58          21   
top     Hyoung-Keun Lee  Banks: Commercial and Savings  Financials   
freq                  1                             51         118   

              location                  website  
count              500                      500  
unique             235                      500  
top     Beijing, China  http://www.macysinc.com  
freq                56                        1

### Series describe

There is also a series.describe() method.  You get different stuff depending on what you try to describe.

#### --> 9.  Series describe()

* Try in one cell for object data: <br />
    `f500['sector'].describe()`
    
* Try in the next for numeric data: <br />
    `f500['Profits_M'].describe()`

Unlike info(),  describe() gives back a Pandas series of its own.  You can assign it to a variable if you need to.

Describe is very useful.  It gives you a quick view into the nature of that column.  

* It can tell you if the data needs to be normalized.
* It can tell you what the "shape" of the data is.


In [34]:
#  object data describe for series 'sector'
f500['sector'].describe()

count            500
unique            21
top       Financials
freq             118
Name: sector, dtype: object

For example, in the below case, the mean is 3055 million in profit. But the median 50% is 1761 million.  The 75th percentile is 3954, nearly the mean.  So this tells you that most of the companies - maybe 70% of them -  made less than average profit.  So 25% of the top Fortune 500 made half of the profit.


In [35]:
# numeric data describe for series 'Profits_M'
f500['Profits_M'].describe()

count      499.000000
mean      3055.203206
std       5171.981071
min     -13038.000000
25%        556.950000
50%       1761.600000
75%       3954.000000
max      45687.000000
Name: Profits_M, dtype: float64

## Aggregate functions to peek at data

Here is a list of the aggregate functions - they also work on a df as well as a series.  They're great for peeking at data to see if it makes sense. 

* series.min() and df.min()
* series.max() and df.max()
* series.mean() and df.mean()
* series.median() and df.median()
* series.mode() and df.mode()
* series.sum() and df.sum()

With a df, you have to tell it the "axis" - what to do the aggregation on 
* to sum each column across rows,  axis = 0 or axis = 'index'.
* to sum each row across columns, axis = 1 or axis = 'column'

For example, to compute the average profit and revenue across all the companies:

#### --> 10.  Aggregation of means via axes

Try:<br />
    `f500[ ['Profits_M','Revenue_M'] ].mean(axis=0)]`

The default axis is 0, actually.  But you may want to compute the mean of all the number columns…

Try:<br />
    `f500.mean(numeric_only=True)`


In [36]:
# mean of profits and revenue #axis=0 sums up all rows, which gives you the column total
f500[['Profits_M','Revenue_M']].mean(axis=0)

Profits_M     3055.203206
Revenue_M    55416.358000
dtype: float64

In [37]:
# mean of all numeric fields only
f500.mean(numeric_only=True)

rank                        250.500000
rank_prev                   222.134000
Revenue_M                 55416.358000
Revenue_Change                4.538353
Profits_M                  3055.203206
Assets_M                 243632.274000
Profit_Change                24.152752
Employees                133998.316000
Years_on_list                15.036000
Stock_Equity              30628.076000
profit_as_pct_revenue         5.652130
profit_as_pct_assets          3.115606
profit_as_pct_equity         16.684221
dtype: float64

## Histogram Value Counts

Pandas provides a ton of tools for analyzing and rearranging data.  One common thing is to create a histogram - a count of the number of occurrence of each particular value in a set of categories.

#### --> 11.  Select sectors
How many companies are in each sector?  First, select out the "sector" series.

**Double Click Here for answer**
<!--
sectors =   f500['sector']
-->

In [38]:
#Setup.  Select out the sector series
sectors =  f500['sector'] # remember this pulls out a single column, not a row.
sectors

title
Walmart                                        Retailing
State Grid                                        Energy
Sinopec Group                                     Energy
China National Petroleum                          Energy
Toyota Motor                      Motor Vehicles & Parts
                                           ...          
Teva Pharmaceutical Industries               Health Care
New China Life Insurance                      Financials
Wm. Morrison Supermarkets             Food & Drug Stores
TUI                                    Business Services
AutoNation                                     Retailing
Name: sector, Length: 500, dtype: object

You can see that there are lots of "Retailing" and "Energy" companies.

`value_counts()` method counts the numbers of each kind of value in a series.


Note that this only works if your data in that series is discrete and from a fixed, clean set.

    What if you had some "Food & Drug" that had '&', with others "Food and Drug" that used "and"?  Or some uppercase, others lowercase?  We ought to clean that up because they would show up as unique, when they're not really unique.


#### --> 12.  Get a count of all the distinct values.

Try:  Use the value_counts() function on the sectors series to get a histogram count of all of the unique values.

**Double Click Here for answer**
<!--
sectors.value_counts()
-->


In [40]:
#  value counts of sectors 
sectors.value_counts()


Financials                       118
Energy                            80
Technology                        44
Motor Vehicles & Parts            34
Wholesalers                       28
Health Care                       27
Food & Drug Stores                20
Transportation                    19
Telecommunications                18
Retailing                         17
Materials                         16
Food, Beverages & Tobacco         16
Industrials                       15
Aerospace & Defense               14
Engineering & Construction        13
Chemicals                          7
Business Services                  3
Hotels, Restaurants & Leisure      3
Household Products                 3
Media                              3
Apparel                            2
Name: sector, dtype: int64

In [41]:
f500['sector'].describe()


count            500
unique            21
top       Financials
freq             118
Name: sector, dtype: object

# Manipulation with Pandas

###  Vector Operations in Pandas

Pandas is an extension of Numpy, so you can do elementwise or broadcasted *vector* add, subtract, multiply, divide on dataframes and series.

#### --> 12. Fun with sums and format strings

First select out the 'profits_M' series from f500 dataset.  Multiply that entire series by 1000000

Then use format strings and `max` to print the profit of the company that earned max profit, with zero decimal places.

THen use format strings and `max` to print the sum of all the profits in trillions.  (Divide millions by a million to get trillions.  A trillion is a million millions.)


**double click here for answer**
<!--
profits_M = f500["profits_M"]
profits * 1000000

# the part of the format after the colon, e.g. :.0f , is a format specifier
#  f is for floating point numbers.  That will show with zero decimals.
print( f"Max profit of f500 companies: ${profits.max():.0f} million.")
# show minimum field width of 6, with at least 2 decimals always.
print( f"Sum of profits: ${f500['Profits_M'].sum()/1000000:6.2f} trillion.")
-->


In [55]:
# Write your code here
profits_M = # select out the series.

profits = profits_M * 1000000   # vector broadcasting of multiplication

# you can leave this example here, or remove it.
x = 123.4
print(f"This example prints a floating point number in a field of 10 chars with 2 decimals: '{x:10.2f}'")
print(f"This example prints a floating point number with zero decimal places:  '{x:.0f}'")




This example prints a floating point number in a field of 10 chars with 2 decimals: '    123.40
This example prints a floating point number with zero decimal places:  '123'


## Set and List Comprehension and selection from Pandas DataFrames

Let's learn about:
* Set comprehensions
* list comprehensions and 
* filters.

You can make a conditional set comprehension just like a list comprehension.  

#### --> 13.  conditional comprehensions and series

* Use a comprehension with `{ }` instead of `[ ]` to pull out the **set** of sectors that have an `&` in them.
* It'll be a conditional comprehension  ` { `*expression` for x in  `*iterator*` if `*condition*` }` 
* Use the `in` operator in the `if` condition part too.


**Double Click Here for answer**
<!--
 { x for x in f500['sector'] if '&' in x }
-->

Output:

```
    {'Aerospace & Defense',
     'Engineering & Construction',
     'Food & Drug Stores',
     'Food, Beverages & Tobacco',
     'Hotels, Restaurants & Leisure',
     'Motor Vehicles & Parts'}
 ```
 

{'Aerospace & Defense',
 'Engineering & Construction',
 'Food & Drug Stores',
 'Food, Beverages & Tobacco',
 'Hotels, Restaurants & Leisure',
 'Motor Vehicles & Parts'}

#### --> 14.  Iterating over a set

After you get that going, assign the set comprehension to a variable `ampersandSectors` and then write a for loop to print them out nicely.

**Double Click Here for answer**
<!--

ampersandSectors = { x for x in f500['sector'] if '&' in x }
for ampSector in ampersandSectors:
    print(ampSector)
    
-->



Output:
```
    Motor Vehicles & Parts
    Food, Beverages & Tobacco
    Engineering & Construction
    Aerospace & Defense
    Food & Drug Stores
    Hotels, Restaurants & Leisure
```

In [27]:
ampersandSectors = # set comprehension of sectors that have '&'
for ???  in  ???:
    print(???)

Motor Vehicles & Parts
Food, Beverages & Tobacco
Engineering & Construction
Aerospace & Defense
Food & Drug Stores
Hotels, Restaurants & Leisure


## A complicated list comprehension

Let's get a tuple of companies and sectors that don't have matching and signs.

This is a fairly complicated expression.  Explore it.  We'll use the principles here quite frequently.  The power of Pandas is in its abiltiy to *explore* data, and these types of statements, along with Jupyter Notebook, make exploring data easier.  As you explore, make notes in Markdown cells!

The `index` property, remember, stores the row labels of our dataframe.

Why generate a tuple?  It's just for checking.... to make sure it worked. We're *exploring*

In [72]:
#Remember the first index on a dataframe, if just one value, is the column. 
#We have to use .loc to select one row.

[ (company,f500.loc[company]['sector']) 
  for company in f500.index 
  if f500.loc[company]['sector'] not in ampersandSectors]


[('Walmart', 'Retailing'),
 ('State Grid', 'Energy'),
 ('Sinopec Group', 'Energy'),
 ('China National Petroleum', 'Energy'),
 ('Royal Dutch Shell', 'Energy'),
 ('Berkshire Hathaway', 'Financials'),
 ('Apple', 'Technology'),
 ('Exxon Mobil', 'Energy'),
 ('McKesson', 'Wholesalers'),
 ('BP', 'Energy'),
 ('UnitedHealth Group', 'Health Care'),
 ('CVS Health', 'Health Care'),
 ('Samsung Electronics', 'Technology'),
 ('Glencore', 'Energy'),
 ('AT&T', 'Telecommunications'),
 ('EXOR Group', 'Financials'),
 ('Industrial & Commercial Bank of China', 'Financials'),
 ('AmerisourceBergen', 'Wholesalers'),
 ('AXA', 'Financials'),
 ('Amazon.com', 'Technology'),
 ('Hon Hai Precision Industry', 'Technology'),
 ('China Construction Bank', 'Financials'),
 ('Total', 'Energy'),
 ('General Electric', 'Industrials'),
 ('Verizon', 'Telecommunications'),
 ('Japan Post Holdings', 'Financials'),
 ('Allianz', 'Financials'),
 ('Cardinal Health', 'Wholesalers'),
 ('Costco', 'Retailing'),
 ('Agricultural Bank of Chin

## DataFrame row filters

To select just certain rows of a dataframe, you use what's called a *filter* which is a boolean-valued series that *marks* certain rows True or False - True if they meet a conditon. False if they don't.

For example This will define a filter for the Health Care companies....

    healthCareFilter = f500['sector'] == 'Health Care'

And this will show which ones matched `Health Care` (are True):

    [ company for company in healthCareFilter.index if healthCareFilter[company] == True ]
    
    
You use the .isin() method on the Series to do the same thing the `in` operator does.  `in` doesn't work because the operator doesn't broadcast across the series. `in` is not a vector operation. `isin()` is a vector function.

Make a filter for all rows where a certain column value is in a set / list / collection of values, with `isin()`:

    ampersandFilter = f500['sector'].isin(ampersandSectors)
  

#### --> 15.  Try those lines below...


In [88]:
healthCareFilter =
ampersandFilter = 

# check healthCareFilter
print([ company for company in healthCareFilter.index if healthCareFilter[company] == True])

print("\nNow check ampersandFilter and peek at the sectors of the selected companies")
[ (company, f500.loc[company]['sector']) 
 for company in ampersandFilter.index 
 if ampersandFilter[company] == True]


['UnitedHealth Group', 'CVS Health', 'Express Scripts Holding', 'Anthem', 'China Resources National', 'Johnson & Johnson', 'Aetna', 'Humana', 'Roche Group', 'Pfizer', 'Bayer', 'Novartis', 'Sinopharm', 'HCA Holdings', 'Sanofi', 'Centene', 'Merck', 'Cigna', 'GlaxoSmithKline', 'Fresenius', 'Gilead Sciences', 'Royal Philips', 'Medtronic', 'AbbVie', 'AstraZeneca', 'Amgen', 'Teva Pharmaceutical Industries']

Now check ampersandFilter and peek at the sectors of the selected companies


[('Toyota Motor', 'Motor Vehicles & Parts'),
 ('Volkswagen', 'Motor Vehicles & Parts'),
 ('Daimler', 'Motor Vehicles & Parts'),
 ('General Motors', 'Motor Vehicles & Parts'),
 ('Ford Motor', 'Motor Vehicles & Parts'),
 ('China State Construction Engineering', 'Engineering & Construction'),
 ('Honda Motor', 'Motor Vehicles & Parts'),
 ('Walgreens Boots Alliance', 'Food & Drug Stores'),
 ('Kroger', 'Food & Drug Stores'),
 ('SAIC Motor', 'Motor Vehicles & Parts'),
 ('Nissan Motor', 'Motor Vehicles & Parts'),
 ('BMW Group', 'Motor Vehicles & Parts'),
 ('China Railway Engineering', 'Engineering & Construction'),
 ('China Railway Construction', 'Engineering & Construction'),
 ('Boeing', 'Aerospace & Defense'),
 ('Nestle', 'Food, Beverages & Tobacco'),
 ('Carrefour', 'Food & Drug Stores'),
 ('Dongfeng Motor', 'Motor Vehicles & Parts'),
 ('Robert Bosch', 'Motor Vehicles & Parts'),
 ('Hyundai Motor', 'Motor Vehicles & Parts'),
 ('Pacific Construction Group', 'Engineering & Construction'),
 ('Te

## Copying a DataFrame, selecting rows
There's a `copy` method on a dataFrame, and it takes a named parameter `deep`, which if you set it to `True` makes sure that the dataFrame copy is a full deep copy, and avoids *references*.

The ampersandFilter has 500 rows.  100 of the rows are True. 

**Note how the sum function sums up just the True rows.  THere's a loose correspondence between 1, 0, True, False.**
* True is oneish.  False is zeroish.
* one is truey.  zero is falsey


In [104]:
f500Copy = f500.copy(deep=True)
f500Copy = f500Copy.loc[ampersandFilter]
print(len(ampersandFilter), len(f500), len(f500Copy), sum(ampersandFilter))

500 500 100 100


See how you can use the filter to get the original rows?  ampersandFilter is *just* a series of True / False values.

In [95]:
f500.loc[ampersandFilter]

rank  rank_prev  Revenue_M  Revenue_Change  \
title                                                                   
Toyota Motor                  5          8     254694             7.7   
Volkswagen                    6          7     240264             1.5   
Daimler                      17         16     169483             2.2   
General Motors               18         20     166380             9.2   
Ford Motor                   21         21     151800             1.5   
...                         ...        ...        ...             ...   
Michelin                    466        451      23120            -1.7   
Heineken Holding            468        459      23044            -0.7   
Sodexo                      484        466      22477            -2.0   
China General Technology    490        383      22113           -20.1   
Wm. Morrison Supermarkets   498        437      21741           -11.3   

                           Profits_M  Assets_M  Profit_Change  Employees  \
title                                                                      
Toyota Motor                 16899.3    437575          -12.3     364445   
Volkswagen                    5937.3    432116            NaN     626715   
Daimler                       9428.4    256262            0.9     282488   
General Motors                9427.0    221690           -2.7     225000   
Ford Motor                    4596.0    237951          -37.7     201000   
...                              ...       ...            ...        ...   
Michelin                      1853.4     26705           43.0     105654   
Heineken Holding               861.5     41469          -18.9      73525   
Sodexo                         707.2     15766          -12.7     425594   
China General Technology       413.6     20860          -20.8      38589   
Wm. Morrison Supermarkets      406.4     11630           20.4      77210   

                                                 ceo  \
title                                                  
Toyota Motor                             Akio Toyoda   
Volkswagen                           Matthias Muller   
Daimler                               Dieter Zetsche   
General Motors                         Mary T. Barra   
Ford Motor                          James P. Hackett   
...                                              ...   
Michelin                       Jean-Dominique Senard   
Heineken Holding           Jean-Francois van Boxmeer   
Sodexo                                 Michel Landel   
China General Technology                 Xu Xianping   
Wm. Morrison Supermarkets             David T. Potts   

                                            industry  \
title                                                  
Toyota Motor                Motor Vehicles and Parts   
Volkswagen                  Motor Vehicles and Parts   
Daimler                     Motor Vehicles and Parts   
General Motors              Motor Vehicles and Parts   
Ford Motor                  Motor Vehicles and Parts   
...                                              ...   
Michelin                    Motor Vehicles and Parts   
Heineken Holding                           Beverages   
Sodexo                                 Food Services   
China General Technology   Engineering, Construction   
Wm. Morrison Supermarkets       Food and Drug Stores   

                                                  sector  \
title                                                      
Toyota Motor                      Motor Vehicles & Parts   
Volkswagen                        Motor Vehicles & Parts   
Daimler                           Motor Vehicles & Parts   
General Motors                    Motor Vehicles & Parts   
Ford Motor                        Motor Vehicles & Parts   
...                                                  ...   
Michelin                          Motor Vehicles & Parts   
Heineken Holding               Food, Beverages & Tobacco   
Sodexo                     Hot

## Using a filter to assign values to specific rows, columns
This one is HUGE.  You'll use it all the time.

This one line will set all values in the "sector" columns for rows that match the filter to a value.


`f500Copy.loc[ampsFilter,"sector"] = "Aerospace and Defense"`

#### --> 16.  Give it a try.


In [110]:
f500Copy = f500.copy(deep=True)  # re-copy to get a clean copy.

airNDefenceFilter = f500Copy['sector'] == 'Aerospace & Defense'

print("Here are the air and defence companies along with their sector.")
for company in f500Copy.index:
    if airNDefenceFilter.loc[company]:
        print(company,  " - ", f500Copy.loc[company]['sector'])
print("\nNow, let's take out the & in one swoop.")


# Take out the & in one assignment.

# Check the result with a for loop - maybe same loop as above.
        

Here are the air and defence companies along with their sector.
Boeing Aerospace & Defense
Airbus Group Aerospace & Defense
China South Industries Group Aerospace & Defense
China North Industries Group Aerospace & Defense
United Technologies Aerospace & Defense
Aviation Industry Corp. of China Aerospace & Defense
Lockheed Martin Aerospace & Defense
China Aerospace Science & Technology Aerospace & Defense
General Dynamics Aerospace & Defense
China Aerospace Science & Industry Aerospace & Defense
China Electronics Technology Group Aerospace & Defense
Northrop Grumman Aerospace & Defense
Raytheon Aerospace & Defense
BAE Systems Aerospace & Defense


## Put it together

See if you can put the pieces together to make a loop to step through the set of all sectors that have an & in them, and then replace the sectors for just those companies with 'and'

You should get something like this - maybe not in this order

#### -->  17.  All together.

```
array(['Retailing', 'Energy', 'Motor Vehicles and Parts', 'Financials',
       'Technology', 'Wholesalers', 'Health Care', 'Telecommunications',
       'Engineering and Construction', 'Industrials',
       'Food and Drug Stores', 'Aerospace and Defense',
       'Food, Beverages and Tobacco', 'Household Products',
       'Transportation', 'Materials', 'Chemicals', 'Media', 'Apparel',
       'Hotels, Restaurants and Leisure', 'Business Services'],
      dtype=object)
```

**Double Click Here for answer**
<!--

f500Copy = f500.copy(deep=True)
ampsectors = { x for x in f500Copy['sector'] if '&' in x }
for ampsector in ampsectors:
    # Build the filter
    ampSectorFilter = f500Copy['sector'] == ampsector
    # use str replace to change & to and
    newSectorName = ampsector.replace('&', 'and')
    # Change the entire dataframe for the filter matches and column
    f500Copy.loc[ampSectorFilter,"sector"] = newSectorName


f500Copy['sector'].unique()

-->


In [113]:
f500Copy = f500.copy(deep=True)
ampsectors = ??? # Set comprehension of all sectors that have & in them
for ampsector in ampsectors:
    # Build the filter
    ampSectorFilter = ???
    # use str replace on ampsector to change & to and
    newSectorName = ???
    # Change the entire dataframe for the filter matches and column
    ??? = newSectorName


# The Pandas Python way is faster.
f500Copy['sector'].unique()   # make a numpy array set out of values in a Pandas series

SyntaxError: invalid syntax (<ipython-input-113-002a5e329a8f>, line 2)

In [114]:
# The python way to generate a set of these.  Pandas way is probably more efficient.

{ x for x in f500Copy['sector'] }

{'Aerospace and Defense',
 'Apparel',
 'Business Services',
 'Chemicals',
 'Energy',
 'Engineering and Construction',
 'Financials',
 'Food and Drug Stores',
 'Food, Beverages and Tobacco',
 'Health Care',
 'Hotels, Restaurants and Leisure',
 'Household Products',
 'Industrials',
 'Materials',
 'Media',
 'Motor Vehicles and Parts',
 'Retailing',
 'Technology',
 'Telecommunications',
 'Transportation',
 'Wholesalers'}

# But why?

In data science, part of your data wrangling will be doing things like:

* Finding columns that have missing data in them, and filling it.
* Finding columns that have *bad* data in them, and striking those rows or replacing the bad data.
* Separating your rows into training, validation, and assessment datasets.
* Looking at relationships between rows and columns and distributions of data to see what makes sense
* Normalizing the data.

Pandas makes that stuff much much easier.  So the skills above will prove vital!

Note that in your Coursera class, all the data was nice and pretty!  It never happens that way in real life.

